### PR TITLE
Draft: POC for upgrading all NPM dependencies, mainly to shake out issues over time

### PR DESCRIPTION
**Pull request by shayneczyzewski**
# Description

This PR is an exploratory proof of concept to see what breaks when we upgrade all of our NPM dependencies to the latest versions.

## Impact
- Wasp's `logout()` behavior changes due to router upgrade [all]
  - Before, when you call it, you'd clear your session _and_ get redirected to the logout path
  - After the upgrade, your session is cleared, but there is no redirect by default
- Users will have to update their usage of `react-router-dom` (e.g. `useHistory` becomes `useNavigate`, params are now `useParams()`, etc.) [`realworld`]
- Users with `axios` declared in their .wasp file will need to upgrade to `^0.27.2` [`waspello`, `waspleau`]
- Users with `@material-ui/core` will need to change to `@mui/material` and update all `require`s [`realworld`]
- Styling via `makeStyles` and `@mui/styles` is not React 18 compatible
- Users cannot use `react-beautiful-dnd@13.1.0` (the package just needs to upgrade their `package.json`, but can `npm install react-beautiful-dnd --save --force` in `.wasp/out/webapp` in meantime) [`waspello`]:
  ```bash
  Web app (stderr): npm ERR! Could not resolve dependency:
  Web app (stderr): npm ERR! peer react@"^16.8.5 || ^17.0.0" from react-beautiful-dnd@13.1.0
  Web app (stderr): npm ERR! node_modules/react-beautiful-dnd
  Web app (stderr): npm ERR!   react-beautiful-dnd@"13.1.0" from the root project
  ```
- Currently, all apps have a warning in the output from CRA bug: https://github.com/facebook/create-react-app/issues/12035 [all]:
  ```bash
  Web app (stderr): (node:42799) [DEP_WEBPACK_DEV_SERVER_ON_AFTER_SETUP_MIDDLEWARE] DeprecationWarning: 'onAfterSetupMiddleware' option is deprecated. Please use the 'setupMiddlewares' option.
  Web app (stderr): (Use `node --trace-deprecation ...` to show where the warning was created)
  Web app (stderr): (node:42799) [DEP_WEBPACK_DEV_SERVER_ON_BEFORE_SETUP_MIDDLEWARE] DeprecationWarning: 'onBeforeSetupMiddleware' option is deprecated. Please use the 'setupMiddlewares' option.
  Web app (stderr): Starting the development server...
   ```
- Prisma has changed some relations requirements (amongst possibly other things) [`realworld`]:
  ```bash
  Db: Error: Get DMMF: Schema parsing - Error while interacting with query-engine-node-api library
  Db (stderr): Error code: P1012
  Db (stderr): error: Error parsing attribute "@relation": Implicit many-to-many relation should not have references argument defined. Either remove it, or change the relation to one-to-many.
  Db (stderr):   -->  schema.prisma:22
  Db (stderr):    |
  Db (stderr): 21 |   favoriteArticles Article[] @relation("FavoritedArticles")
  Db (stderr): 22 |   followedBy User[] @relation("FollowedUser", references: [id])
  Db (stderr): 23 |   following User[] @relation("FollowedUser", references: [id])
  Db (stderr):    |
  ```

Fixes # (issue)

## Type of change

Please select the option(s) that is more relevant.

- [ ] Code cleanup
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

### CODE REVIEW

**Comment by Martinsos** [comment]
@shayneczyzewski nice job! So you updated all the deps, and you also fixed all the issues in our codebase?
Obviously these are breaking changes, but we are ok with that.
Btw we don't necessarily have to update all of them to the latest versions -> we can say smth like "we will not upgrade to React 18 yet because the rest of the ecosystem didn't catch up with it yet". I am making this up, but you get the idea. I would say we want to update to the latest versions that are stable / play well with the rest of the ecosystem. And we do that when we are ok with releasing new version of Wasp with breaking changes (which in Alpha is always ok, later we will want to be a bit more conservative).
So what are the next steps here?

File: waspc/data/Generator/templates/server/src/config.js [hunk 1]
```diff
@@ -35,5 +34,5 @@ const config = {
   }
 }
 
-const resolvedConfig = _.merge(config.all, config[env])
+const resolvedConfig = {...config.all, ...config[env]}
```
Comment (Martinsos) [inline waspc/data/Generator/templates/server/src/config.js:37]: There is a difference here though -> `_.merge` does the merge recursively, while spread operator will merge only at the top level.
So if we have
```js
let foo = { a: { b: 3 } }
let bar = { a: { c: 4 } }
```

`_.merge(foo, bar)` will give `{ a: { b: 3, c: 4 } }` while `{...foo, ...bar}` will give `{ a: { c: 4 } }`.

Maybe we ok with such behaviour, but I just wanted to warn it is a change indeed. Maybe it is safer to go with recursive merge here.

**Review by Martinsos — COMMENTED** [review]
Looking good, there is one possibly tricky change that I commented on (_.merge vs spread operator).
I guess what we would also need to do it some additional changes to the docs / tutorials? Since we updated some of the packages? Also, example of course, but that we can do once we are all ready here.

File: waspc/data/Generator/templates/server/src/config.js (hunk 1 above)
Comment (shayneczyzewski) [reply waspc/data/Generator/templates/server/src/config.js:37]: Yeah, I had thought about that and was trying to find a non-library recursive solution (I didn't even remember I just did the quick and dirty merge to remove lodash), and was tinkering with my own/others' recursive solutions. This def would not go in like this, but I appreciate the note here. 👍🏻 

In general, nothing in this Draft PR would be considered mergable just FYI. Only testing dep upgrades/removals atm. Thanks!

**Comment by shayneczyzewski** [comment]
> @shayneczyzewski nice job! So you updated all the deps, and you also fixed all the issues in our codebase? Obviously these are breaking changes, but we are ok with that. Btw we don't necessarily have to update all of them to the latest versions -> we can say smth like "we will not upgrade to React 18 yet because the rest of the ecosystem didn't catch up with it yet". I am making this up, but you get the idea. I would say we want to update to the latest versions that are stable / play well with the rest of the ecosystem. And we do that when we are ok with releasing new version of Wasp with breaking changes (which in Alpha is always ok, later we will want to be a bit more conservative). So what are the next steps here?

Thanks! Yep, I tried to put everything up to the latest and see/fix what broke. Good point on us being able to punt on React 18. I can go back through and upgrade everything _except_ that (and all the libraries to their latest that are still compatible).

Nothing immediate to do here, this will mainly be a background task for me over next couple weeks just to get some versions bumped but without being too disruptive to users/us. This was just the first attempt to see what _could_ break if we went the latest on everything. I will dial it back some with React now haha 👍🏻

**Comment by Martinsos** [comment]
> > @shayneczyzewski nice job! So you updated all the deps, and you also fixed all the issues in our codebase? Obviously these are breaking changes, but we are ok with that. Btw we don't necessarily have to update all of them to the latest versions -> we can say smth like "we will not upgrade to React 18 yet because the rest of the ecosystem didn't catch up with it yet". I am making this up, but you get the idea. I would say we want to update to the latest versions that are stable / play well with the rest of the ecosystem. And we do that when we are ok with releasing new version of Wasp with breaking changes (which in Alpha is always ok, later we will want to be a bit more conservative). So what are the next steps here?
> 
> Thanks! Yep, I tried to put everything up to the latest and see/fix what broke. Good point on us being able to punt on React 18. I can go back through and upgrade everything _except_ that (and all the libraries to their latest that are still compatible).
> 
> Nothing immediate to do here, this will mainly be a background task for me over next couple weeks just to get some versions bumped but without being too disruptive to users/us. This was just the first attempt to see what _could_ break if we went the latest on everything. I will dial it back some with React now haha 👍🏻

Ok if you think that using older React is indeed better, but I used that as a made up example, I have no idea if using React 18 is early or not!

**Comment by shayneczyzewski** [comment]
> > > @shayneczyzewski nice job! So you updated all the deps, and you also fixed all the issues in our codebase? Obviously these are breaking changes, but we are ok with that. Btw we don't necessarily have to update all of them to the latest versions -> we can say smth like "we will not upgrade to React 18 yet because the rest of the ecosystem didn't catch up with it yet". I am making this up, but you get the idea. I would say we want to update to the latest versions that are stable / play well with the rest of the ecosystem. And we do that when we are ok with releasing new version of Wasp with breaking changes (which in Alpha is always ok, later we will want to be a bit more conservative). So what are the next steps here?
> > 
> > 
> > Thanks! Yep, I tried to put everything up to the latest and see/fix what broke. Good point on us being able to punt on React 18. I can go back through and upgrade everything _except_ that (and all the libraries to their latest that are still compatible).
> > Nothing immediate to do here, this will mainly be a background task for me over next couple weeks just to get some versions bumped but without being too disruptive to users/us. This was just the first attempt to see what _could_ break if we went the latest on everything. I will dial it back some with React now haha 👍🏻
> 
> Ok if you think that using older React is indeed better, but I used that as a made up example, I have no idea if using React 18 is early or not!

I will still try to find some time for all of us to chat on it this week. We can see what others think too. Thanks!

**Comment by shayneczyzewski** [comment]
Closing this one to just focus on React 17 and other smaller, less intrusive updates.

**Pull request closed without merging**